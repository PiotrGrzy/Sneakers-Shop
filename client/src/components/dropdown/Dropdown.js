import React, { useState } from 'react';
import { connect } from 'react-redux';
import { setQuery } from '../../redux/search/search.actions';
import './dropdown.scss';

const Dropdown = ({ type, title, itemsList, setQuery }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [items, setItems] = useState(itemsList);

  const toggleSelect = (id) => {
    // selects listitem by its id
    // finds index
    const index = items.findIndex((item) => item.id === id);
    // clone list
    const updatedItems = [...items];
    // updates item to selected or unselected
    updatedItems[index].selected = !items[index].selected;
    // update local state of selected items
    setItems(updatedItems);
    // update query with new selected items names
    setQuery(
      type,
      updatedItems.filter((item) => item.selected).map((item) => item.name)
    );
  };

  const toggleList = () => {
    // reveal dropdowm
    setIsOpen(!isOpen);
  };

  return (
    <div className="dd-wrapper">
      <div className="dd-header" onClick={toggleList}>
        <div className="dd-header-title">{title}</div>
        {isOpen ? (
          <i className="lni lni-arrow-up"></i>
        ) : (
          <i className="lni lni-arrow-down"></i>
        )}
      </div>
      {isOpen && (
        <ul className="dd-list">
          {items.map(({ id, name, selected }) => (
            <li
              className="dd-list-item"
              key={id}
              onClick={() => toggleSelect(id)}
            >
              {name}{' '}
              {selected && <i className="lni lni-star-filled dd-checkmark"></i>}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default connect(null, { setQuery })(Dropdown);
