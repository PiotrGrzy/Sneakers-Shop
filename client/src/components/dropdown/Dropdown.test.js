import React from 'react';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import { Dropdown } from './Dropdown';

const props = {
  type: 'test',
  title: 'Test',
  itemsList: [
    { id: 0, name: 'item 1', selected: false },
    { id: 1, name: 'item 2', selected: true },
    { id: 2, name: 'item 3', selected: false },
  ],
  setQuery: jest.fn(),
};

describe('Dropdown', () => {
  const wrapper = shallow(<Dropdown {...props} />);
  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  describe('hides and reveal options on click', () => {
    const wrapper = mount(<Dropdown {...props} />);
    it('renders all options on click', () => {
      wrapper.find('.dd-header').simulate('click');
      expect(wrapper.find('.dd-list-item').length).toEqual(3);
    });

    it('mark selected option with the "star" icon', () => {
      wrapper.find('.dd-list-item').at(0).simulate('click');
      expect(wrapper.find('.dd-list-item').at(0).childAt(1).exists()).toBe(
        true
      );
    });

    it('hides all options on second click', () => {
      wrapper.find('.dd-header').simulate('click');
      expect(wrapper.find('.dd-list-item').length).toEqual(0);
    });
  });
});
