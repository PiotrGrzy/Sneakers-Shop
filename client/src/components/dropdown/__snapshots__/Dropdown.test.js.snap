// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders properly 1`] = `ShallowWrapper {}`;
